File: commons-parsers/src/main/java/xyz/wagyourtail/commons/parsers/impl/constant/number/DecimalPart.java
package xyz.wagyourtail.commons.parsers.impl.constant.number;

import lombok.val;
import xyz.wagyourtail.commons.core.reader.CharReader;
import xyz.wagyourtail.commons.core.reader.StringCharReader;
import xyz.wagyourtail.commons.parsers.Data;
import xyz.wagyourtail.commons.parsers.StringData;

public class DecimalPart extends StringData.OnlyRaw<Data.SingleContent<String>> {

    private DecimalPart(String content) {
        super(content, DecimalPart::getContentChecked);
    }

    public DecimalPart(CharReader<?> content) {
        super(content, DecimalPart::getContentChecked);
    }

    public static DecimalPart parse(String rawContent) {
        StringCharReader reader = new StringCharReader(rawContent);
        val number = new DecimalPart(reader);
        reader.expectEOS();
        return number;
    }

    public static DecimalPart unchecked(String rawContent) {
        return new DecimalPart(rawContent);
    }

    private static SingleContent<String> getContentChecked(CharReader<?> reader) {
        val first = reader.peek();
        if (first < '0' || first > '9') {
            throw reader.createException("Not a decimal number: " + (char) first);
        }

        return new SingleContent<>(reader.takeWhile(e -> e >= '0' && e <= '9'));
    }
}

File: commons-parsers-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/parsers/impl/constant/number/DecimalPart.kt
package xyz.wagyourtail.commonskt.parsers.impl.constant.number

import xyz.wagyourtail.commonskt.parsers.Data
import xyz.wagyourtail.commonskt.parsers.StringData
import xyz.wagyourtail.commonskt.reader.CharReader

class DecimalPart(rawContent: String) : StringData.OnlyRaw<Data.SingleContent<String>>(rawContent) {

    constructor(content: SingleContent<String>) : this(content.toString())

    companion object : StringDataBuilder<DecimalPart, SingleContent<String>> {

        override fun invoke(rawContent: CharReader<*>): DecimalPart {
            return DecimalPart(checkedBuildContent(rawContent))
        }

        override fun checkedBuildContent(reader: CharReader<*>): SingleContent<String> {
            val first = reader.peek()
            if (first == null || !first.isDigit()) {
                throw reader.createException("Not a decimal number: $first")
            }
            return SingleContent(reader.takeWhile { it.isDigit() })
        }

    }

    override fun checkedBuildContent(reader: CharReader<*>): SingleContent<String> {
        return Companion.checkedBuildContent(reader)
    }


}